### ISSUE TITLE
low IOU and large variation of mAP

### ISSUE THREAD

**Issue by smartyfn** [AlexeyAB/darknet#2552]
Thanks so much for such as an excellent Yolo model.  We are using yolov3-tiny.cfg and yolo3-tiny.conv.15 weights to train 5 class of abnormal cells.  For the first 4 classes, most bounding boxes contain single cell, but for 5th class, bounding boxes contain a cluster of cells.  The anchors were calculated below.  The avg IOU is low, but mAP is relative stable.
My cfg file:[yolov3-tiny.cfg.txt](https://github.com/AlexeyAB/darknet/files/2940426/yolov3-tiny.cfg.txt)

![anchor](https://user-images.githubusercontent.com/44492549/53944901-4def2d00-40fb-11e9-9982-47cd786cf5b5.jpg)
![map](https://user-images.githubusercontent.com/44492549/53945659-08336400-40fd-11e9-9cdd-b7509c120164.jpg)

Some detected boxes were shifted due to low IOU, I assume.  I’ve modified all the annotation for the 5th class to ”4 0.5 0.5 1 1” to include entire 416x416 image and recalculate anchors. We get anchors of 46, 43,   56, 57,   86, 83,  142,140,  233,231,  415,414 with avg IOU of 82.46%.
However, I have noticed a large variation of mAP, particularly for the 5th class (2% to 95%, just as from %2 to 95%, then from 95% to 10%), as shown below:
![map11](https://user-images.githubusercontent.com/44492549/53946094-fb634000-40fd-11e9-8421-30d5d3ee43bd.jpg)
My cfg file:[yolov3-tiny-second.cfg.txt](https://github.com/AlexeyAB/darknet/files/2940490/yolov3-tiny-second.cfg.txt)

My questions are:
1. What causes such a large mAP variation?
2. Should we use modified model to train both small and large objects?
3. Since the cluster size could exceeds 416x416, should you use size of 608 or 832 to train?
4. Why IOU is so low (70.63%) before I modify the annotation file and how can be improved?

**Comment by AlexeyAB** [comment]
@smartyfn Hi,

> I’ve modified all the annotation for the 5th class to ”4 0.5 0.5 1 1” to include entire 416x416 image and recalculate anchors. 

What is the real **average** size of 5th class?

1. Usually, mutually exclusive markings lead to this (for example, in some pictures buses are marked as class 1, on others buses are marked as class 2, and in the third buses are not marked), etc.

May be in your case, since you set 1x1 relative size of objects, then **during data augmentation** some part of object will be cropped (because some part of object will be out of image), and size of object will be reduced - so that the corners of the object are not outside the image: https://github.com/AlexeyAB/darknet/blob/b751bac17505a742f149ada81d75689b5e692cde/src/data.c#L245-L248
So the same object will have different size depend on what part of this object is out of image. This may introduce a small amount of uncertainty. But in this case, strong fluctuations of mAP should decrease, the longer you train. **So just try to train 200 000 - 300 000 iterations instead of 100 000. Or don't set 1x1 relative size of objects.**


2. Do you mean should you use yolo-tiny_3l.cfg? I think no, because you don't have very small objects (the smallest anchors 46x43). You can try to use `yolov3.cfg` if your GPU is powerfull enough.

3. No. If you set ”4 0.5 0.5 1 1” for all 5th class object, then 

4. `IOU is so low (70.63%)` - it is just IoU of anchors, not of detections. To get IoU of detection, you should run `./darknet detector map`... command: https://github.com/AlexeyAB/darknet#when-should-i-stop-training
To increase IoU of anchors, just calculate more anchors, for example 15, ...
`./darknet detector calc_anchors data/obj.data -num_of_clusters 15 -width 416 -height 416`
But then you should set new anchors, num=15, and new masks=... for each yolo-layer: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

**Comment by smartyfn** [comment]
@AlexeyAB Thanks for your quick answer. 

> What is the real average size of 5th class?
Usually, mutually exclusive markings lead to this (for example, in some pictures buses are marked as class 1, on others buses are marked as class 2, and in the third buses are not marked), etc.
May be in your case, since you set 1x1 relative size of objects, then during data augmentation some part of object will be cropped (because some part of object will be out of image), and size of object will be reduced - so that the corners of the object are not outside the image: 

What does "1x1 relative size of objects" mean?  In my case both train_image_width and detection_image_width are 416. Our picture has size of 40000 x 40000，the real average size of 5th class is very big which looks like a circular ring. I crop some part of the 5th class as 416 x 416 randomly. So the size of the 5th class varies from half to full dimension in the 416x416 image.

**Comment by AlexeyAB** [comment]
@smartyfn 

> What does "1x1 relative size of objects" mean?

It means width=1 height=1 in label txt-file, as you set:

>  I’ve modified all the annotation for the 5th class to ”4 0.5 0.5 1 1” to include entire 416x416 image and recalculate anchors. 

----

> Our picture has size of 40000 x 40000，the real average size of 5th class is very big which looks like a circular ring. I crop some part of the 5th class as 416 x 416 randomly. So the size of the 5th class varies from half to full dimension in the 416x416 image.

* Did you crop 416x416 image from 40000 x 40000 image?

----

>  I’ve modified all the annotation for the 5th class to ”4 0.5 0.5 1 1” to include entire 416x416 image and recalculate anchors. 


* Did you just change all anotation to `4 0.5 0.5 1 1` for 5th class even if it has completely different size?
* Did you do it for validation dataset too? Or do you use `valid=train.txt` in obj.data file?
* Do you want to detect the exact coordinates of objects or accuracy you just need to know whether the object in the 416x416 image or not?

I think this mutually exclusive labels (on one image object at 0.3,0.3 but is labeled as 0.5,0.5, on another image object at 0.8,0.8 but is labeled as 0.5,0.5) lead to large mAP variation.

----

* How many training images do you have?

Also it seems that mAP increasing isn't stopped for `yolov3-tiny.cfg.txt`, so you can try to train it more than 100 000 iterations.

**Comment by smartyfn** [comment]
@AlexeyAB You are  warmhearted. Thank you very much.

> Do you mean should you use yolo-tiny_3l.cfg? I think no, because you don't have very small objects (the smallest anchors 46x43). You can try to use yolov3.cfg if your GPU is powerfull enough.

If I want to improve accuracy . Should I use yolov3.conv.81 or darknet53.conv.74 ?
> Did you crop 416x416 image from 40000 x 40000 image?

Yes. All the 5 classes of pictures are cropped from 40000 x 40000 image, and  the bounding boxes of the 5th class is bigger than other three classes

> Did you just change all anotation to 4 0.5 0.5 1 1 for 5th class even if it has completely different size?

Yes, the 5th class  has completely different sizes which vary from 290 x 290 to  416 x 416

> Did you do it for validation dataset too? Or do you use valid=train.txt in obj.data file?

I use valid = valid.txt. I use 10% of the total dataset as validation dataset, don't include negative images

> Do you want to detect the exact coordinates of objects or accuracy you just need to know whether the object in the 416x416 image or not?

I just need to detect the exact coordinates of objects or accuracy in the 416x416 image.

> I think this mutually exclusive labels (on one image object at 0.3,0.3 but is labeled as 0.5,0.5, on another image object at 0.8,0.8 but is labeled as 0.5,0.5) lead to large mAP variation.

I think so. I doubt that the 5th class mark has a problem

> How many training images do you have?

Our dataset has 18 thousand pictures to train and 2 thousand pictures to validate, because we don't have many original pictures, so some pictures are created by rotated and flipped, besides we added 13700 of  negative images to improve accuracy

Because our image is cell image instead of regular image like person,car,dog etc. How could we do to improve accuracy obviously ?

**Comment by smartyfn** [comment]
Can I  use size of 608 or 832 to train in my case . I think it will save much time to analyze the entire image of 40000 x 40000

**Comment by AlexeyAB** [comment]
@smartyfn 

1. If your dataset looks more like MS COCO images then use `yolov3.conv.81`, if it looks more like ImageNet images then use `darknet53.conv.74` for training `yolov3.cfg` **(I would recommend to use `yolov3-spp.cfg`)**. I think you should use `darknet53.conv.74`.
If you train yolov3-tiny.cfg then use `yolov3-tiny.conv.15` pre-trained file: https://github.com/AlexeyAB/darknet#how-to-train-tiny-yolo-to-detect-your-custom-objects

2. Yes, you can train your network with `width=608 height=608` in cfg-file and with 608x608 images (which are cropped from 40000x40000).
If you trained your model with `width=416 height=416` in cfg-file, then you can change `width=608 height=608` for Detection only if you Trained it with `random=1`.
In both these cases you can faster analyze 40000x40000 image.

----

> I just need to detect the exact coordinates of objects or accuracy in the 416x416 image.

If you want to detect real exact coordinates of objects on 416x416 images, then you should use real coordinates and sizes of objects rather than `0.5 0.5 1 1` in Training dataset

----

> Our dataset has 18 thousand pictures to train and 2 thousand pictures to validate, because we don't have many original pictures, so some pictures are created by rotated and flipped, besides we added 13700 of negative images to improve accuracy

So, in total, do you have 18+14 ~= 32 000 images (416x416)?

3. To increase accuracy

* use `yolov3-spp.cfg` instead of yolov3-tiny.cfg

* increase training dataset by adding new labeled images

* increase your dataset by using data augmentation (rotataion, ...)

* train more iterations, until mAP increase stop

* It is usually easier to find a large dataset for classification than for detection, since for classification it is easier to label images (you need to specify only the class of the object, but you do not need to specify the coordinates). So, **if you can, try to find very large dataset of cells for classification, 100 000 - 10 000 000 images that are labeled for classification only** (for example, each jpg-filename contains class_name or class_id that is present on this image, for example img_`n000`.jpg for class=1, img_`n001`.jpg for class=1, ...). Then train classifier 
`./darknet classifier train cfg/classification.data cfg/darknet53_448.cfg darknet53.conv.74  -topk`
where classification.data contains:
```
classes=1000
train  = data/train.txt
valid  = data/train.txt
backup = backup
labels = data/labels_in_jpg_filenames.txt
names  = data/object.names
top=5
```
`data/labels_in_jpg_filenames.txt` (n001, n002, ...) and `data/object.names` (dog, cat, ...) should have the same number of lines.

After training extract pre-trained weights darknet53.conv.74
`./darknet partial cfg/darknet53_448.cfg darknet53_448.weights darknet53.conv.74 74`

And then train your detector `yolov3-spp.cfg` by using `darknet53.conv.74`
`./darknet detector train data/obj.data yolov3-spp_obj.cfg darknet53.conv.74 -map`

**Comment by smartyfn** [comment]
@AlexeyAB Thanks for your good suggestions.

> So, in total, do you have 18+14 ~= 32 000 images (416x416)?
Yes

**Comment by AlexeyAB** [comment]
@smartyfn 

* https://github.com/pjreddie/darknet/blob/master/cfg/darknet53_448.cfg
* https://pjreddie.com/media/files/darknet53_448.weights

More: https://pjreddie.com/darknet/imagenet/
